### PR TITLE
Use `xparse` commands in new commands/environments tutorial

### PR DESCRIPTION
**Pull request by marcin-serwin**
This PR changes the the tutorial on new commands and environments to use new standard based on `xparse`.

### CODE REVIEW

File: book/src/custom.tex [hunk 1]
```diff
@@ -22,19 +22,22 @@ \chapter{Customising \LaTeX}
...
 \section{New Commands, Environments and Packages}
 
-You may have noticed that all the commands I introduce in this
-book are typeset in a box, and that they show up in the index at the end
-of the book. Instead of directly using the necessary \LaTeX{} commands
-to achieve this, I have created a \wi{package} in which I defined new
-commands and environments for this purpose. Now I can simply write:
+At the beginning of this book we have mentioned that \LaTeX{} allows us to
+write documents using logical markup, such as \ci{emph} or \ci{section}. There
+may be however situations where a command to produce the logical unit of your
```
Comment (oetiker) [inline book/src/custom.tex:27]:
```suggestion
may be however situations where \LaTeX{} does not provide an appropriate command for the content you want write about```

File: book/src/custom.tex [hunk 2]
```diff
@@ -22,19 +22,22 @@ \chapter{Customising \LaTeX}
...
+document is not provided by \LaTeX{}. You may have noticed that all the
```
Comment (oetiker) [inline book/src/custom.tex:28]:
```suggestion
You may have noticed that all the
```

File: book/src/custom.tex [hunk 3]
```diff
@@ -22,19 +22,22 @@ \chapter{Customising \LaTeX}
...
+commands I introduce in this book are typeset in a box, and that they show up
+in the index at the end of the book. There is no \LaTeX{} markup for element
+``this is a command introduction'', bu \LaTeX{} allows me to define one. Thus I
+can write:
```
Comment (oetiker) [inline book/src/custom.tex:32]:
```suggestion
commands I introduce in this book are typeset in a box, and that they show up
in the index at the end of the book. There is no \LaTeX{} markup to format example code or commands, but \LaTeX{} allows me to define new commands. Thus I
can write:
```

File: book/src/custom.tex [hunk 4]
```diff
@@ -219,29 +405,51 @@ \subsection{Your Own Package}
...
 \end{lscommand}
-\noindent for use at the very beginning of your package
-file. \verb|\ProvidesPackage| tells \LaTeX{} the name of the package
-and will allow it to issue a sensible error message when you try to
-include a package twice. Figure~\ref{package} shows a small example
-package that contains the commands defined in the examples above.
+\noindent for use at the very beginning of your package file. This command
+switches the \LaTeX{} to process the  file in \emph{expl mode}. There are some
+differences in the way this mode is handled. The most visible change is the
+fact that whitespaces are ignored. You may have noticed that in a lot of
```
Comment (oetiker) [inline book/src/custom.tex:438]:
```suggestion
fact that all whitespace is ignored. You may have noticed that in a lot of
```

**Review by oetiker — COMMENTED** [review]
cool

File: book/src/custom.tex [hunk 5]
```diff
@@ -22,19 +22,21 @@ \chapter{Customising \LaTeX}
 
 \section{New Commands, Environments and Packages}
 
-You may have noticed that all the commands I introduce in this
-book are typeset in a box, and that they show up in the index at the end
-of the book. Instead of directly using the necessary \LaTeX{} commands
-to achieve this, I have created a \wi{package} in which I defined new
-commands and environments for this purpose. Now I can simply write:
+At the beginning of this book we have mentioned that \LaTeX{} allows us to
+write documents using logical markup, such as \ci{emph} or \ci{section}. There
```
Comment (oetiker) [inline book/src/custom.tex:26]:
```suggestion
write documents using logical markup, with commands like \ci{emph} or \ci{section}. There
```

File: book/src/custom.tex [hunk 6]
```diff
@@ -22,19 +22,21 @@ \chapter{Customising \LaTeX}
...
+may be however situations where \LaTeX{} does not provide an appropriate command for the content you want write about.```
+You may have noticed that all the
+commands I introduce in this book are typeset in a box, and that they show up
+in the index at the end of the book. There is no \LaTeX{} markup to format example code or commands, but \LaTeX{} allows me to define new commands. Thus I
```
Comment (oetiker) [inline book/src/custom.tex:30]:
```suggestion
in the index at the end of the book. There is no \LaTeX{} markup to format example code or commands, but \LaTeX{} allows me to define my own commands for this purpose. Thus I
```

File: book/src/custom.tex [hunk 7]
```diff
@@ -54,134 +56,315 @@ \subsection{New Commands} \label{sec:new_commands}
...
-If missing it defaults to 0, i.e. no argument allowed.
-
-The following two examples should help you to get the idea.
-The first example defines a new command called \ci{tnss}. This is
+\noindent command. It requires three arguments: the \carg{name} of the command
+you want to create, the \carg{arg}ument \carg{spec}ification and the
+\emph{definition} of the command.
+
+The \carg{argspec} argument specifies the number and types of arguments the
+command receives. Two most important types are \cargv{m} and \cargv{o} which
```
Comment (oetiker) [inline book/src/custom.tex:66]:
```suggestion
command receives. The two most important types are \cargv{m} and \cargv{o} which
```

File: book/src/custom.tex [hunk 8]
```diff
@@ -54,134 +56,315 @@ \subsection{New Commands} \label{sec:new_commands}
...
+stand for, respectively, mandatory and optional. Thus if you want to create a
```
Comment (oetiker) [inline book/src/custom.tex:67]:
```suggestion
command receives. The two most important types are \cargv{m}, for \emph{mandatory} and \cargv{o} for \emph{optional}. To create a
```

File: book/src/custom.tex [hunk 9]
```diff
@@ -54,134 +56,315 @@ \subsection{New Commands} \label{sec:new_commands}
...
+
+% in the document body:
+\begin{itemize}
+\item \txsit{long}
+\item \txsit*{long}
+\end{itemize}
+\end{example}
+
+These are just the most common argument specifications. For a full overview of
+those take a look at the documentation of \pai*{xparse} package.\footnote{This
```
Comment (oetiker) [inline book/src/custom.tex:191]:
```suggestion
These are just the most common argument specifications. For a full description of
take a look at the \pai*{xparse} package documentation.\footnote{This
```

File: book/src/custom.tex [hunk 10]
```diff
@@ -54,134 +56,315 @@ \subsection{New Commands} \label{sec:new_commands}
...
+package is included by default in new \LaTeX{} versions, so you probably do not
+need to include it manually.}
```
Comment (oetiker) [inline book/src/custom.tex:193]:
```suggestion
package is part of the current \LaTeX{} release, so you probably do not
need to install it manually.}
```

Comment (marcin-serwin) [reply book/src/custom.tex:193]: This comment wasn't about installing but about doing `\usepackage{xparse}`, hence the word include. Do you think other word would be more appropriate?

Comment (oetiker) [reply book/src/custom.tex]: oops :)

Comment (oetiker) [reply book/src/custom.tex]: how about this wording ?

File: book/src/custom.tex [hunk 11]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+    No optional argument given.%
+  }%
+}
+
+\MyCommand\\
+\MyCommand[hello]
+\end{example}
+
+There are two variations of it: \ci{IfValueT} and \ci{IfValueF} which may
+be used if you only need output for one of the branches. Thus the example with
```
Comment (oetiker) [inline book/src/custom.tex:132]:
```suggestion
be used if you only need output for one of the branches. The example with
```

File: book/src/custom.tex [hunk 12]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+  \IfValueT{#1}{\emph{#1} }%
+  #2 Introduction to \LaTeXe}
+
+% in the document body:
+\begin{itemize}
+\item \txsit{definitive}
+\item \txsit[very]{long}
+\end{itemize}
+\end{example}
+There are also commands \ci{IfNoValueTF}, \ci{IfNoValueT} and \ci{IfNoValueF}
```
Comment (oetiker) [inline book/src/custom.tex:146]:
```suggestion
The commands \ci{IfNoValueTF}, \ci{IfNoValueT} and \ci{IfNoValueF}
```

File: book/src/custom.tex [hunk 13]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+which work exactly the same, but the value\slash no-value branches are swapped.
```
Comment (oetiker) [inline book/src/custom.tex:147]:
```suggestion
work exactly the same, but the value\slash no-value branches are swapped.
```

File: book/src/custom.tex [hunk 14]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+
+Often you will want to use optional arguments when present but use some default
+values when the user does not provide them. This could be achieved by
+\ci{IfValueTF}, but a much better approach exists: \cargv{O} argument
```
Comment (oetiker) [inline book/src/custom.tex:151]:
```suggestion
\ci{IfValueTF}, but with the \cargv{O} argument a default value can be set directly.
```

File: book/src/custom.tex [hunk 15]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+specification. It works like \cargv{o} but allows us to specify default value
```
Comment (oetiker) [inline book/src/custom.tex:152]:
```suggestion
It works like \cargv{o} but allows us to specify default value
```

File: book/src/custom.tex [hunk 16]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+if no value is supplied. Thus we can write
```
Comment (oetiker) [inline book/src/custom.tex:153]:
```suggestion
if no value is supplied. Write
```

File: book/src/custom.tex [hunk 17]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+
+% in the document body:
+\begin{itemize}
+\item \txsit{definitive}
+\item \txsit[very]{long}
+\end{itemize}
+\end{example}
+
+Another useful argument specification is \cargv{s}, short for star. This
+argument enables us to provide different definitions based on whether the
```
Comment (oetiker) [inline book/src/custom.tex:168]:
```suggestion
argument allows to provide different definitions based on whether the
```

File: book/src/custom.tex [hunk 18]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+\begin{itemize}
+\item \txsit{long}
+\item \txsit*{long}
+\end{itemize}
+\end{example}
+
+These are just the most common argument specifications. For a full description,
+take a look at the \pai*{xparse} package documentation.\footnote{This
+package is loaded automatically in new \LaTeX{} releases, so you probably do not
+need to put \texttt{\textbackslash usepackage\{xparse\}} in your preamble. }
```
Comment (oetiker) [inline book/src/custom.tex:192]: perfect

File: book/src/custom.tex [hunk 19]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
 already defined, \LaTeXe{} will silently ignore it.
 
-There are some points to note about whitespace following \LaTeX{} commands. See
-page \pageref{whitespace} for more information.
-
 \subsection{New Environments}
-Just as with the \verb|\newcommand| command, there is a command
-to create your own environments. The \ci{newenvironment} command uses the
+Just as with the \ci{NewDocumentCommand} command, there is a command to create
+your own environments. The \ci{NewDocumentEnvironment} command uses the
```
Comment (oetiker) [inline book/src/custom.tex:206]:
```suggestion
The \ci{NewDocumentEnvironment} command lets you create your own environments. It uses the
```

File: book/src/custom.tex [hunk 20]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
-When creating a new environment you may easily get bitten by extra spaces
-creeping in, which can potentially have fatal effects, for example when you
-want to create a title environment which supresses its own indentation as
-well as the one on the following paragraph. The \ci{ignorespaces} command in
-the begin block of the environment will make it ignore any space after
-executing the begin block. The end block is a bit more tricky as special
-processing occurs at the end of an environment. With the
-\ci{ignorespacesafterend} \LaTeX{} will issue an \ci{ignorespaces} after the
-special `end' processing has occurred.
+Note that when environment argument are read they are read \emph{after} the
```
Comment (oetiker) [inline book/src/custom.tex:236]:
```suggestion
Note that when environment argument are read, they are read \emph{after} the
```

File: book/src/custom.tex [hunk 21]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+  Second time:\\
+  #1
+} {}
+
+\begin{twice}
+This will be printed twice!
+\end{twice}
+\end{example}
+While this makes one of the \carg{at begin}, \carg{at end} arguments redundant,
+they are still required. (In the example above we provided empty \carg{at
```
Comment (oetiker) [inline book/src/custom.tex:293]:
```suggestion
they are still required. (In the example above we provided an empty \carg{at
```

File: book/src/custom.tex [hunk 22]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+
+To understand why this happens it is instructive to consider how \TeX{} expands
+the defined commands. The above \ci{RenewDocumentCommand} tells the \TeX{}
+engine that whenever \ci*{emph}[foo:vm] is seen it must replace it with
+\verb|\emph{foo}~(``foo'' is emphasised)|. You may already see the problem here.
+In the next stage it will again replace the \ci*{emph}[foo:vm] yielding
+\begin{verbatim}
+\emph{foo}~(``foo'' is emphasised)~(``foo'' is emphasised)
+\end{verbatim}
+It is easy to see that this process will never end and at some point \TeX{}
```
Comment (oetiker) [inline book/src/custom.tex:327]:
```suggestion
This process will never end and at some point \TeX{}
```

File: book/src/custom.tex [hunk 23]
```diff
@@ -54,134 +56,314 @@ \subsection{New Commands} \label{sec:new_commands}
...
+will suffer the same fate since \verb|\oldemph{...}| will be replaced by \TeX{}
+with \verb|\emph{...}| and the cycle continues.
+
+In order to avoid this problem a special command exists
+\begin{lscommand}
+  \ci*{NewCommandCopy}[name:m ! command:m]
+\end{lscommand}
+It makes the \carg{name} the exact copy of the \carg{command}. You may
+appropriate the difference between this and \ci{NewDocumentCommand} in the
+following example
```
Comment (oetiker) [inline book/src/custom.tex:346]:
```suggestion
It makes the \carg{name} the exact copy of the \carg{command}. The following example shows how this works```

File: book/src/custom.tex [hunk 24]
```diff
@@ -219,29 +403,54 @@ \subsection{Your Own Package}
...
+\ci*{ProvidesExplPackage}[package name:m ! date:m ! version:m ! description:m]
 \end{lscommand}
-\noindent for use at the very beginning of your package
-file. \verb|\ProvidesPackage| tells \LaTeX{} the name of the package
-and will allow it to issue a sensible error message when you try to
-include a package twice. Figure~\ref{package} shows a small example
-package that contains the commands defined in the examples above.
+\noindent for use at the very beginning of your package file. This command
+switches the \LaTeX{} to process the  file in \emph{expl mode}. There are some
+differences in the way this mode is handled. The most visible change is the
```
Comment (oetiker) [inline book/src/custom.tex:435]:
```suggestion
\noindent for use at the very beginning of your package file. This command
tells the \LaTeX{} to process the  file in \emph{expl mode}. The most visible effect of this is
```

File: book/src/custom.tex [hunk 25]
```diff
@@ -219,29 +403,54 @@ \subsection{Your Own Package}
...
+fact that all whitespace is ignored. You may have noticed that in a lot of
```
Comment (oetiker) [inline book/src/custom.tex:436]:
```suggestion
that all whitespace is ignored. You may have noticed that in many of the
```

File: book/src/custom.tex [hunk 26]
```diff
@@ -219,29 +403,54 @@ \subsection{Your Own Package}
...
+examples above we had to end most lines with \ai{\%} in order to get correct
```
Comment (oetiker) [inline book/src/custom.tex:437]:
```suggestion
examples above we had to end most lines with \ai{\%} to get correct
```

File: book/src/custom.tex [hunk 27]
```diff
@@ -219,29 +403,54 @@ \subsection{Your Own Package}
...
+spacing. This change forces us to insert spaces explicitly as they are usually
```
Comment (oetiker) [inline book/src/custom.tex:438]:
```suggestion
spacing in the output. In \emph{expl mode}, spaces have to be added explicitly if needed at all. They are usually
```

File: book/src/custom.tex [hunk 28]
```diff
@@ -219,29 +403,54 @@ \subsection{Your Own Package}
...
+much rarer when writing a package. In order to insert spaces you have to use
```
Comment (oetiker) [inline book/src/custom.tex:439]:
```suggestion
quite rare when writing a package. To insert spaces, use
```

File: book/src/custom.tex [hunk 29]
```diff
@@ -219,29 +403,54 @@ \subsection{Your Own Package}
...
+the \ai{\~} character, which normally denotes non-breaking space.\footnote{If
+  you want to insert non-breaking space in the \emph{expl mode} you have to use
```
Comment (oetiker) [inline book/src/custom.tex:441]:
```suggestion
  you want to insert non-breaking space in \emph{expl mode}, use
```

File: book/src/custom.tex [hunk 30]
```diff
@@ -219,29 +403,54 @@ \subsection{Your Own Package}
...
+  \ci{nobreakspace}.} Paragraphs can be started with the \ci{par} command.
+
+The arguments are used to provide information about package in the log file. If
+you used this package and looked at the log file you would find
```
Comment (oetiker) [inline book/src/custom.tex:445]:
```suggestion
you use this package and look at the log file you will find
```

**Review by oetiker — COMMENTED** [review]
a few more ideas
